daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -89,156 +97,190 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
-                  s"Mismatching variant id, the type tells us $typeVariantId, but the value tells us $id"
-                )
-            )
-            val pkg = unsafeGetPackage(typeVariantId.packageId)
-            val (dataTypParams, variantDef) =
-              assertRight(SignatureLookup.lookupVariant(pkg, typeVariantId.qualifiedName))
-            variantDef.constructorRank.get(constructorName) match {
-              case None =>
-                fail(
-                  s"Couldn't find provided variant constructor $constructorName in variant $typeVariantId"
+        def typeMismatch = fail(s"mismatching type: $ty and value: $value")
+        val (ty1, tyArgs) = destructApp(ty0)
+        ty1 match {
+          case TBuiltin(bt) =>
+            tyArgs match {
+              case Nil =>
+                (bt, value) match {
+                  case (BTUnit, ValueUnit) =>
+                    SValue.SUnit
+                  case (BTBool, ValueBool(b)) =>
+                    if (b) SValue.SValue.True else SValue.SValue.False
+                  case (BTInt64, ValueInt64(i)) =>
+                    SValue.SInt64(i)
+                  case (BTTimestamp, ValueTimestamp(t)) =>
+                    SValue.STimestamp(t)
+                  case (BTDate, ValueDate(t)) =>
+                    SValue.SDate(t)
+                  case (BTText, ValueText(t)) =>
+                    SValue.SText(t)
+                  case (BTParty, ValueParty(p)) =>
+                    SValue.SParty(p)
+                  case _ =>
+                    typeMismatch
+                }
+              case typeArg0 :: Nil =>
+                (bt, value) match {
+                  case (BTNumeric, ValueNumeric(d)) =>
+                    typeArg0 match {
+                      case TNat(s) =>
+                        Numeric.fromBigDecimal(s, d).fold(fail, SValue.SNumeric)
+                      case _ =>
+                        typeMismatch
+                    }
+                  case (BTContractId, ValueContractId(c)) =>
+                    cids += c
+                    SValue.SContractId(c)
+                  case (BTOptional, ValueOptional(mbValue)) =>
+                    mbValue match {
+                      case Some(v) =>
+                        SValue.SOptional(Some(go(typeArg0, v, newNesting)))
+                      case None =>
+                        SValue.SValue.None
+                    }
+                  case (BTList, ValueList(ls)) =>
+                    if (ls.isEmpty) {
+                      SValue.SValue.EmptyList
+                    } else {
+                      SValue.SList(ls.map(go(typeArg0, _, newNesting)))
+                    }
+                  // textMap
+                  case (BTTextMap, ValueTextMap(entries)) =>
+                    if (entries.isEmpty) {
+                      SValue.SValue.EmptyTextMap
+                    } else {
+                      SValue.SGenMap(
+                        isTextMap = false,
```

Comment:
I think you replaced the wrong one? you are matching on `TextMap` here so this should be true. The one below where you’re matching on `GenMap` on the other hand should be false.

Reply:
I need another coffee.

Reply:
I had a test.

Reply:
who needs coffee if you have tests instead :slightly_smiling_face: